capsules/extra/src/ieee802154/phy_driver.rs

```diff
@@ -0,0 +1,460 @@
+                    .and_then(|write| {
+                        write.enter(|payload| {
+                            let frame_len = payload.len();
+                            let dst_start = hil::radio::PSDU_OFFSET;
+                            let dst_end = dst_start + frame_len;
+                            payload.copy_to_slice(&mut kbuf[dst_start..dst_end]);
+
+                            self.radio.transmit(kbuf, frame_len).map_or_else(
+                                |(errorcode, error_buf)| {
+                                    self.kernel_tx.replace(error_buf);
+                                    Err(errorcode)
+                                },
+                                |()| {
+                                    self.current_app.set(processid);
+                                    Ok(())
+                                },
+                            )
+                        })
+                    })?
+            })
+        })?
+    }
+
+    /// If the driver is currently idle and there are pending transmissions,
+    /// pick an app with a pending transmission and return its `ProcessId`.
+    fn get_next_tx_if_idle(&self) -> Option<ProcessId> {
+        if self.current_app.is_some() {
+            return None;
+        }
+        let mut pending_app = None;
+        for app in self.apps.iter() {
+            let processid = app.processid();
+            app.enter(|app, _| {
+                if app.pending_tx {
+                    pending_app = Some(processid);
+                }
+            });
+            if pending_app.is_some() {
+                break;
+            }
+        }
+        pending_app
+    }
+
+    /// Schedule the next transmission if there is one pending.
+    fn do_next_tx(&self) {
+        self.get_next_tx_if_idle()
+            .map(|processid| match self.perform_tx(processid) {
+                Ok(()) => {}
+                Err(e) => {
+                    let _ = self.apps.enter(processid, |_app, upcalls| {
+                        let _ = upcalls.schedule_upcall(
+                            upcall::FRAME_TRANSMITTED,
+                            (kernel::errorcode::into_statuscode(Err(e)), 0, 0),
+                        );
+                    });
+                }
+            });
+    }
+}
+
+impl<'a, R: hil::radio::Radio<'a>> SyscallDriver for RadioDriver<'a, R> {
+    /// IEEE 802.15.4 low-level control.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Driver existence check.
+    /// - `1`: Return radio status. Ok(())/OFF = on/off.
+    /// - `2`: Set short address.
+    /// - `4`: Set PAN ID.
+    /// - `5`: Set channel.
+    /// - `6`: Set transmission power.
+    /// - `7`: Commit any configuration changes.
+    /// - `8`: Get the short MAC address.
+    /// - `10`: Get the PAN ID.
+    /// - `11`: Get the channel.
+    /// - `12`: Get the transmission power.
+    /// - `27`: Transmit a frame. The frame must be stored in the write RO allow
```

Comment:
Isn't `must be the length of the frame` problematic for MPU requirements? I.e., what if the frame is 87 bytes or something? I think the buffer has to be a power of two sometimes. Wouldn't it be better to pass in a length, store it in `app.pending_tx = Some(arg1)` and take the minimum of the buffer length and provided length when transmitting?

Reply:
Why does the MPU matter? The buffer is in the app's RAM. Are you thinking of IPC?

Reply:
Oops... never mind. Yes. I still think having an explicit length is better, but no matter.